cachix/src/Cachix/Client/Commands.hs

```diff
@@ -150,16 +150,20 @@ use env name useOptions = do
       addBinaryCache (config env) binaryCache useOptions
         $ fromMaybe (getInstallationMode nixEnv) (useMode useOptions)
 
--- TODO: lots of room for perfomance improvements
+-- TODO: lots of room for performance improvements
 push :: Env -> PushArguments -> IO ()
 push env (PushPaths opts name rawPaths) = do
   hasNoStdin <- hIsTerminalDevice stdin
-  when (not hasNoStdin && not (null rawPaths)) $ throwIO $ AmbiguousInput "You provided both stdin and store path arguments, pick only one to proceed."
-  inputStorePaths <-
+  stdinContents <-
     if hasNoStdin
-      then return rawPaths
-      else T.words <$> getContents
-  -- TODO: if empty, take whole nix store and warn: nix store-path --all
+      then return ""
+      else getContents
+  when (stdinContents /= "" && not (null rawPaths)) $ throwIO
+    $ AmbiguousInput "You provided both stdin and store path cli arguments, provide only one to proceed."
+  let inputStorePaths =
+        if stdinContents == ""
+          then rawPaths
+          else T.words stdinContents
```

Comment:
stdinContents handling introduces a corner case (what if `stdinContents == " "`) and isn't very clear. Ideas:
 - immediately `T.words` stdinContents
 - turn three conditionals into one by pattern matching on `(stdinContents, rawPaths) :: ([Text], [Text])`

It's also a bit weird that `rawPaths` magically turn into `inputStorePaths` which aren't raw. Probably naming can be simplified.